added object.uniformValues for overwriting material.uniforms

arose opened the pull request:
Continuing from #6021. Allow a new property `.uniformValues` in renderable objects that overwrites values in `material.uniforms`.

arose committed:
added object.uniformValues for overwriting material.uniforms

WestLangley commented on examples/webgl_interactive_instances_gpu.html:
```diff
@@ -0,0 +1,443 @@
+
+		<script src="../build/three.min.js"></script>
+
+		<script src="../src/renderers/WebGLRenderer.js"></script>
```
remove WebGLRenderer source

WestLangley commented on src/renderers/WebGLRenderer.js:
```diff
@@ -4564,6 +4572,25 @@ THREE.WebGLRenderer = function ( parameters ) {
+
+			for ( var j = 0, jl = o_values.length; j < jl; j ++ ) {
+
+				u = o_values[ j ];
```
`u` is not declared.

WestLangley commented:
Nice demo!

If you have a look at `renderer.info` in the console, you will see that the number of programs on the GPU side is not affected by this feature -- the shader program is shared anyway. In light of that, what is the benefit of this PR in your view?

Also, as you change your menu items, `renderer.info` shows the number of geometries increasing. Is there a memory leak of some sort?

WestLangley commented:
I'm seeing some bugs in which the material color depends on which object is rendered first.

Here is the pattern I am using:

	mesh = new THREE.Mesh( geometry, material );
	mesh.position.set( - 10, 0, 0 );
	scene.add( mesh );

	mesh = new THREE.Mesh( geometry, material );
	mesh.uniformValues = [
		[ "diffuse", new THREE.Color( 0x00ffff ) ]
	];
	mesh.position.set( 10, 0, 0 );
	scene.add( mesh );

arose committed:
only temporarily overwrite uniforms with values from object

arose committed:
some simplifications

arose commented:
Thanks @WestLangley  for having a look at it!

> remove WebGLRenderer source

I'll keep that in mind for when the PR is ready

> `u` is not declared.

Fixed.

> If you have a look at `renderer.info` in the console, you will see that the number of programs on the GPU side is not affected by this feature -- the shader program is shared anyway. In light of that, what is the benefit of this PR in your view?

It is an optimization to avoid the additional time needed for initializing separate materials for each mesh.

> Also, as you change your menu items, `renderer.info` shows the number of geometries increasing. Is there a memory leak of some sort?

Fixed, I re-created the highlight-box geometry without disposing of it.

> I'm seeing some bugs in which the material color depends on which object is rendered first.

Oh, I had not anticipated that one would use the same material with objects of which some have `.uniformValues` and some not. But, that is a valid usage and is now supported -- the material uniform values are now only temporarily overwritten.

I guess refreshing the uniform values could be made more fine grained, that is, load only those to the GPU that where actually overwritten.

WestLangley commented:
This is interesting.  I would be nice if, for example, a collection of sprites could share the same sprite sheet, and same material, while having overridden offset/repeat values.

Or, as you said, if all objects could share the same uber material, and have overridden material uniforms -- pushing to the GPU only those overridden values.

arose committed:
attempt at more fine-grained uniforms GPU upload

arose commented:
Did an attempt at more fine-grained uniforms GPU upload. This seems to increase rendering speed.

mrdoob commented:
New approach: #7048
